src/rgw/rgw_sal_d3n.cc

```diff
@@ -92,129 +92,96 @@ int D3NFilterObject::D3NFilterReadOp::prepare(optional_yield y, const DoutPrefix
-      r = d->flush(std::move(completed));
-      if (r < 0) {
-        lsubdout(g_ceph_context, rgw, 0) << "D3NFilterObject::get_obj_iterate_cb:: " << __func__ << "(): Error: failed to drain/flush, r= " << r << dendl;
-      }
-      return r;
+    const std::lock_guard l(d3n_get_data.d3n_lock);
+    auto oid = completed.front().obj.get_ref().obj.oid;
+    if (bl.length() <= g_conf()->rgw_get_obj_max_req_size) {
+      ldpp_dout(dpp, 20) << "D3NFilterObject: " << __func__ << "(): bl.length <= rgw_get_obj_max_req_size (default 4MB) - write to datacache, bl.length=" << bl.length() << dendl;
+      filter->get_d3n_cache()->put(bl, bl.length(), oid);
     } else {
-      // Write To Cache
-      ldpp_dout(dpp, 20) << "D3NFilterObject::get_obj_iterate_cb:: " << __func__ << "(): WRITE TO CACHE: oid=" << read_obj.oid << ", obj-ofs=" << obj_ofs << ", read_ofs=" << read_ofs << " len=" << len << dendl;
-      auto completed = d->aio->get(obj, rgw::Aio::librados_op(std::move(op), d->yield), cost, id);
-      return d->flush(std::move(completed));
+      ldpp_dout(dpp, 20) << "D3NFilterObject: " << __func__ << "(): not writing to datacache - bl.length > rgw_get_obj_max_req_size (default 4MB), bl.length=" << bl.length() << dendl;
     }
+    completed.pop_front_and_dispose(std::default_delete<rgw::AioResultEntry>{});
   }
-  ldpp_dout(dpp, 20) << "D3NFilterObject::get_obj_iterate_cb:: " << __func__ << "(): Warning: Check head object cache handling flow, oid=" << read_obj.oid << dendl;
-
   return 0;
 }
 
-
 int D3NFilterObject::D3NFilterReadOp::iterate(const DoutPrefixProvider* dpp, int64_t ofs, int64_t end,
 			RGWGetDataCB* cb, optional_yield y)
 {
-  rgw::sal::Driver* driver = filter->get_next();
-  rgw::sal::RadosStore* rados_store = static_cast<rgw::sal::RadosStore*>(driver);
-  RGWRados* rados = rados_store->getRados();
-  CephContext *cct = rados->ctx();
-  const uint64_t chunk_size = cct->_conf->rgw_get_obj_max_req_size;
-  const uint64_t window_size = cct->_conf->rgw_get_obj_window_size;
-
-  rctx = std::make_unique<RGWObjectCtx>(driver);
-  auto aio = rgw::make_throttle(window_size, y);
-  get_obj_data data(rados, cb, &*aio, ofs, y);
-  get_obj_priv_data priv_data(&data, filter);
+  const uint64_t window_size = g_conf()->rgw_get_obj_window_size;
+  std::string oid = source->get_key().get_oid();
+  off_t len = (end - ofs) + 1;
+  const uint64_t cost = len;
+  const uint64_t id = ofs;
   
-  int r = rados->iterate_obj(dpp, *rctx, source->get_bucket()->get_info(),
-			                      source, ofs, end, chunk_size, get_obj_iterate_cb, &priv_data, y);
+  ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << "ofs: " << ofs << "end: " << end << ", len: " << len << dendl;
   
-  if (r < 0) {
-    ldpp_dout(dpp, 0) << "iterate_obj() failed with " << r << dendl;
-    data.cancel();
+  client_cb = cb;
+  aio = rgw::make_throttle(window_size, y);
+  RGWSI_RADOS::Obj obj;
```

Comment:
@cbodley : I attempted a hacky way to use Aio here, but I get a seg fault here, for reasons that I understand:

1146	  io_ctx_impl->get();
(gdb) bt
#0  librados::v14_2_0::IoCtx::operator= (this=this@entry=0x55aed4aed3a0, rhs=...) at /home/prsrivas/ceph/src/librados/librados_cxx.cc:1146
#1  0x000055aecb35ffb2 in RGWSI_RADOS::Pool::State::operator= (this=0x55aed4aed3a0) at /home/prsrivas/ceph/src/rgw/services/svc_rados.h:91
#2  RGWSI_RADOS::Pool::operator= (this=this@entry=0x55aed4aed358) at /home/prsrivas/ceph/src/rgw/services/svc_rados.h:83
#3  0x000055aecb360003 in RGWSI_RADOS::rados_ref::operator= (this=0x55aed4aed358) at /home/prsrivas/ceph/src/rgw/services/svc_rados.h:149
#4  0x000055aecb35f4d5 in RGWSI_RADOS::Obj::operator= (this=<optimized out>) at /home/prsrivas/ceph/src/rgw/services/svc_rados.h:154
#5  rgw::YieldingAioThrottle::get(RGWSI_RADOS::Obj 

So what is the alternative here, probably I did not understand what you said in the meeting.

Also, I can see that a file gets written to the cache, after it is fetched from the store. But the name of the file doesn't have the object instance prepended to it. I tried concatenating the instance id with the oid, and to get the oid of the key (done above), but I do not get the instance id at all. How can I get that? (This is just an attempt to figure out things for a 4MBsized object, I will add multi-part related code once I deal with this simple scenario)

Reply:
> @cbodley : I attempted a hacky way to use Aio here, but I get a seg fault here, for reasons that I understand:

yeah it looks like it's crashing because Obj is uninitialized

> So what is the alternative here, probably I did not understand what you said in the meeting.

i was thinking that we would only need the Aio stuff in the rados backend, but had forgotten that `d3n_cache_op()` depended on it too. i think what you've done here is mostly fine, it's just the `RGWSI_RADOS::Obj` part we need to get rid of. i'll propose a change to the Aio interface for you and @adamemerson to review

> Also, I can see that a file gets written to the cache, after it is fetched from the store. But the name of the file doesn't have the object instance prepended to it. I tried concatenating the instance id with the oid, and to get the oid of the key (done above), but I do not get the instance id at all. How can I get that?

[rgw_obj_key::get_oid()](https://github.com/ceph/ceph/blob/7f65724/src/rgw/rgw_obj_types.h#L238-L254) should append the instance id if there is one; are you testing on a versioned bucket?

Reply:
> i'll propose a change to the Aio interface for you and @adamemerson to review

see https://github.com/ceph/ceph/pull/50347

Reply:
> > @cbodley : I attempted a hacky way to use Aio here, but I get a seg fault here, for reasons that I understand:
> 
> yeah it looks like it's crashing because Obj is uninitialized
> 
> > So what is the alternative here, probably I did not understand what you said in the meeting.
> 
> i was thinking that we would only need the Aio stuff in the rados backend, but had forgotten that `d3n_cache_op()` depended on it too. i think what you've done here is mostly fine, it's just the `RGWSI_RADOS::Obj` part we need to get rid of. i'll propose a change to the Aio interface for you and @adamemerson to review
> 
> > Also, I can see that a file gets written to the cache, after it is fetched from the store. But the name of the file doesn't have the object instance prepended to it. I tried concatenating the instance id with the oid, and to get the oid of the key (done above), but I do not get the instance id at all. How can I get that?
> 
> [rgw_obj_key::get_oid()](https://github.com/ceph/ceph/blob/7f65724/src/rgw/rgw_obj_types.h#L238-L254) should append the instance id if there is one; are you testing on a versioned bucket?

@cbodley : i am not testing on a versioned bucket, but when I was testing on the previous version of this PR for the same bucket I saw a alphanumeric string prefixed to the oid, I presumed that is the instance id, (correct me if I am wrong) and now with these latest changes I dont see that prefix - its just the name of the object that I uploaded.

Reply:
@pritha-srivastava that was probably from the namespace? we use that `ns` field to hide multipart parts and tail (shadow) objects

Reply:
Not sure if that was 'ns', but even if we are uploading objects to a non-versioned bucket, shouldn't each object have an instance id associated?

Reply:
rgw_obj_key::instance should be empty for objects in a non-versioned bucket

Reply:
@cbodley: What is the prefix in 'cc6703c0-bcce-46e0-a910-fd995ec8624b.4147.1_4M.txt'. This file is added to the datacache when I try to get a 4MB object which is named 4M.txt. This file is added to the cache when my latest commit (to decouple rados) isnt part of the branch.

Reply:
ah, that prefix is coming from `rgw_bucket::marker`. i forget exactly where that gets added to the object name

Reply:
> > i'll propose a change to the Aio interface for you and @adamemerson to review
> 
> see #50347

@cbodley : I used #50347, and was able to get a clean run for uploading and getting a 4 MB object, with the file getting written to the cache. I will focus on multipart objects and on improving this PR. I haven't used rgw_bucket::marker for storing the object in the cache yet, but I think I will eventually need to figure this out too.